File: 02_Ros2Basics/03_Ros2Services/01_ServiceClient.py
# import the empty module from std_servs Service interface
from std_srvs.srv import Empty
# import the ROS2 Python client libraries
import rclpy
from rclpy.node import Node


class ClientAsync(Node):

    def __init__(self):
        # Here you have the class constructor

        # call the class constructor to initialize the node as service_client
        super().__init__('service_client')
        # create the Service Client object
        # defines the name and type of the Service Server you will work with.
        """This is the line where you create the Client, that uses the Empty Service type and connects to a Service named /moving. """
        self.client = self.create_client(Empty, 'moving')
        # checks once per second if a Service matching the type and name of the Client is available.
        """This while loop is used to ensure that the Service Server (in this case, /moving) is up and running"""
        while not self.client.wait_for_service(timeout_sec=1.0):
            # if it is not available, a message is displayed
            self.get_logger().info('service not available, waiting again...')
        
        # create an Empty request
        self.req = Empty.Request()
        

    def send_request(self):
        
        # send the request
        """Send an asynchronous request to the Service Server using the call_async() method. 
        Then, store the response from the Server in the variable self.future. 
        After making the request, the Server will immediately return future, 
        which indicates whether the call and response are finished (but it does not 
        contain the value of the response itself)"""
        self.future = self.client.call_async(self.req)


def main(args=None):
    # initialize the ROS communication
    rclpy.init(args=args)
    # declare the node constructor
    client = ClientAsync()
    # run the send_request() method
    client.send_request()

    while rclpy.ok():
        # pause the program execution, waits for a request to kill the node (ctrl+c)

        """This will spin the client node one time. It is similar to rclpy.spin (that you already saw ), but instead of spinning the node indefinitely, it will only spin it once."""
        rclpy.spin_once(client)
        #you keep checking (while the program is running, rclpy.ok()) if the Service response is finished:
        if client.future.done():
            try:
                # checks the future for a response from the Service
                # while the system is running. 
                # If the Service has sent a response, the result will be written
                # to a log message.
                # If it is finished, you then get the value of the Server response
                response = client.future.result()
            except Exception as e:
                # Display the message on the console
                client.get_logger().info(
                    'Service call failed %r' % (e,))
            else:
                # Display the message on the console
                # While you wait for the Service response to finish, you are printing messages to the node's log:
                client.get_logger().info(
                    'the robot is moving' ) 
            break

    client.destroy_node()
    # shutdown the ROS communication
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: 04_Ros2Basics_cpp/04_Services/01_Service_Client.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/empty.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>

using namespace std::chrono_literals;

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("service_client");
  
  //This is the line where you create the Client:
  //You can see that you create a Client that uses the Empty Service type and connects to a Service named /stop
  rclcpp::Client<std_srvs::srv::Empty>::SharedPtr client =
    node->create_client<std_srvs::srv::Empty>("stop");

  auto request = std::make_shared<std_srvs::srv::Empty::Request>();

    //You can see that you create a Client that uses the Empty Service type and connects to a Service named /moving
  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "service not available, waiting again...");
  }

/*Send an asynchronous request to the Service Server using the async_send_request() method. 
Then, store the response from the Server in the variable result_future. 
This result_future variable will contains what is known as a future object. 
After making the request, the Server will immediately return result_future, 
which indicates whether the call and response are finished (but it does not
 contain the value of the response itself). You will read more about 
 this in the next section of this chapter.*/

  auto result_future = client->async_send_request(request);
  // Wait for the result.
//   /Finally, you spin the node until this result_future is completed (the service is completed):
  if (rclcpp::spin_until_future_complete(node, result_future) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
  //If it is finished successfully, you then get the value of the Server response and print a message indicating that the robot is moving:
    auto result = result_future.get();
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "The robot is moving");
  } else {
  //If it is finished successfully, you then get the value of the Server response and print a message indicating that the robot is moving:
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Failed to call service /moving");
  }

  rclcpp::shutdown();
  return 0;
}
